tools/c-parser/syntax_transforms.ML

```diff
@@ -4,6 +4,131 @@
  * SPDX-License-Identifier: BSD-2-Clause
  *)
 
+signature FUN_RM =
+sig
+  (* Statement lists. *)
+  type stmt_list
+  val stmt_list_of : Absyn.statement list -> stmt_list
+  val of_stmt_list : stmt_list -> Absyn.statement list
+  val @@ : stmt_list * stmt_list -> stmt_list
+  val @- : stmt_list * Absyn.statement -> stmt_list
+  val @> : stmt_list * Absyn.statement list -> Absyn.statement list
+  val empty : stmt_list -> bool
+  (* A monad for extracting function calls that are embedded in expressions. *)
+  type 'rv fun_rm
+  val run : 'a fun_rm -> stmt_list * 'a
+  val run' : unit fun_rm -> Absyn.statement list
+  val scope : 'a fun_rm -> (stmt_list * 'a) fun_rm
+  val scope' : unit fun_rm -> Absyn.statement list fun_rm
+  val add_stmts : stmt_list -> unit fun_rm
+  val add_stmt : Absyn.statement -> unit fun_rm
+  val next_sym : string -> int fun_rm
+  val return : 'a -> 'a fun_rm
+  val >- : 'a fun_rm * ('a -> 'b fun_rm) -> 'b fun_rm
+  val <:> : 'a fun_rm * 'b fun_rm -> ('a * 'b) fun_rm
+  val >> : unit fun_rm * 'b fun_rm -> 'b fun_rm
+  val >>- : 'a fun_rm * ('a -> unit fun_rm) -> 'a fun_rm
+  val <$ : ('a -> 'b) * 'a fun_rm -> 'b fun_rm
+  val $> : 'a fun_rm * ('a -> 'b) -> 'b fun_rm
+  val mmap : ('a -> 'b fun_rm) -> 'a list -> 'b list fun_rm
+end
+
+structure FunRm : FUN_RM =
+struct
+
+infix >- <*> <$ $> <:> >> >>-
+
+fun (f >- g) (syms, stmts) = let
+  val (syms', stmts', rv_f) = f (syms, stmts)
+in
+  g rv_f (syms', stmts')
+end
+
+fun return v (syms, stmts) = (syms, stmts, v)
+fun f <*> g = f >- (fn f' => g >- return o f')
+fun f <$ g = return f <*> g
+fun f $> g = g <$ f
+fun f <:> g = pair <$ f <*> g
+fun f >> g = snd <$ (f <:> g)
+fun f >>- g = f >- (fn x => g x >> return x)
+
+fun mmap _ [] = return []
+  | mmap f (h::t) = cons <$ f h <*> mmap f t
+
+(* A simple difference list. *)
```

Comment:
As we discussed on Mmost earlier: while the code will rarely see lists long enough to make naive concatenation suboptimal, I'm happy for the difflist to remain in place. However, given the use of ` option`, one should replace the highlighted comment with something more descriptive.